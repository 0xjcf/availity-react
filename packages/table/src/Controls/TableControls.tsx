import React, { cloneElement, isValidElement, useEffect, useState } from 'react';
import { Form } from 'reactstrap';

type Props = {
  id?: string;
  disabled?: boolean;
  children?: React.ReactNode[] | React.ReactNode;
} & React.HTMLAttributes<HTMLElement>;

const TableControls = ({ id, disabled, children, ...rest }: Props): JSX.Element => {
  const [isDisabled, setIsDisabled] = useState<boolean>(disabled || false);

  useEffect(() => {
    setIsDisabled(disabled || false);
  }, [disabled]);

  const childElements = Array.isArray(children) ? children : [children];
  return (
    <div id={id} className="table-controls" {...rest}>
      <Form inline>
        {childElements?.map((child: React.ReactNode, index: number) => {
          if (!isValidElement(child)) {
            return null;
          }

          return cloneElement(child, { disabled: isDisabled, key: `table_controls_${index}` });
        })}
      </Form>
    </div>
  );
};

export default TableControls;
