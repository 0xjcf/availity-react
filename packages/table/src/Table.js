import React from 'react';
import 'core-js/stable';
import PropTypes from 'prop-types';
import { useTable, useRowSelect, useSortBy } from 'react-table';
import { Util, Table as RsTable } from 'reactstrap';
import Icon from '@availity/icon';
import TableHeader from './TableHeader';
import TableHeaderRow from './TableHeaderRow';
import TableHeaderCell from './TableHeaderCell';
import IndeterminateCheckbox from './IndeterminateCheckbox';
import TableRow from './TableRow';
import TableCell from './TableCell';
import { TableContext } from './TableContext';

const Table = ({
  id,
  additionalContent: AdditionalContent,
  bodyProps,
  columns,
  cellProps,
  headerProps,
  onRowClick,
  onRowSelected,
  records,
  rowProps,
  scrollable,
  selectable,
  sortable,
  initialState,
  ...rest
}) => {
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
    selectedFlatRows: selectedRows,
    allColumns,
    toggleAllRowsSelected,
  } = useTable(
    {
      columns,
      data: records,
      initialState: initialState || {},
    },
    useSortBy,
    useRowSelect,
    (hooks) => {
      if (selectable) {
        hooks.visibleColumns.push((columns) => [
          {
            id: 'selection',
            title: 'Select record(s)',
            className: 'fixed-width-selection',
            defaultCanSort: false,
            disableClick: true,
            Header: ({ getToggleAllRowsSelectedProps }) => {
              const selectedProps = {
                onClick: async () => {
                  await toggleAllRowsSelected();
                  if (onRowSelected) {
                    onRowSelected({ selectedRows });
                  }
                },
              };

              return (
                <div className="text-center">
                  <IndeterminateCheckbox
                    data-testid="table_header_select_all"
                    aria-label="Select all records"
                    {...selectedProps}
                    {...getToggleAllRowsSelectedProps()}
                  />
                </div>
              );
            },
            Cell: ({ row: { original, toggleRowSelected, getToggleRowSelectedProps, index } }) => {
              const selectedProps = {
                onClick: () => {
                  if (onRowSelected) {
                    toggleRowSelected();
                    onRowSelected({ selectedId: original.id, data: original });
                  }
                },
              };

              return (
                <div className="text-center">
                  <IndeterminateCheckbox
                    data-testid={`table_header_select_row_${index}`}
                    aria-label="Select record"
                    {...selectedProps}
                    {...getToggleRowSelectedProps()}
                  />
                </div>
              );
            },
          },
          ...columns,
        ]);
      }

      if (!sortable) {
        hooks.visibleColumns.forEach((col) => {
          col.disableSortBy = true;
        });
      }
    }
  );

  let handleRowClick;
  let handleCellClick;

  if (selectable && onRowClick) {
    handleCellClick = onRowClick;
  } else {
    handleRowClick = onRowClick;
  }

  const populateId = () => (id ? `${id}_` : '');

  return (
    <TableContext.Provider
      value={{
        allColumns,
        selectedRows,
        scrollable,
        AdditionalContent,
      }}
    >
      <RsTable id={id} {...getTableProps({ className: 'av-grid' })} {...rest}>
        <TableHeader id={`${populateId()}table_header`} {...headerProps}>
          {headerGroups.map((headerGroup, rowIndex) => (
            <TableHeaderRow
              id={`${populateId()}table_header_row_${rowIndex}`}
              data-testid={`${populateId()}table_header_row_${rowIndex}`}
              key={rowIndex.toString()}
              headerGroup={headerGroup}
            >
              {headerGroup.headers.map((column, cellIndex) => (
                <TableHeaderCell
                  id={`${populateId()}table_header_row_${rowIndex}_cell_${cellIndex}_${column.id}`}
                  data-testid={`${populateId()}table_header_row_${rowIndex}_cell_${cellIndex}_${column.id}`}
                  key={column.id}
                  column={column}
                >
                  {column.render('Header')}
                  {sortable && column.defaultCanSort ? (
                    <Icon
                      aria-hidden="true"
                      name={column.isSorted ? (column.isSortedDesc ? 'sort-down' : 'sort-up') : 'sort'}
                    />
                  ) : null}
                </TableHeaderCell>
              ))}
            </TableHeaderRow>
          ))}
        </TableHeader>
        <tbody {...getTableBodyProps()} {...bodyProps}>
          {rows.map((row, rowIndex) => {
            prepareRow(row);
            return (
              <TableRow
                id={`${populateId()}table_row_${rowIndex}`}
                data-testid={`${populateId()}table_row_${rowIndex}`}
                key={row.original.id}
                index={rowIndex}
                row={row}
                onRowClick={handleRowClick}
                onCellClick={handleCellClick}
                {...rowProps}
              >
                {row.cells.map((cell, cellIndex) => (
                  <TableCell
                    id={`${populateId()}table_row_${rowIndex}_cell_${cellIndex}`}
                    data-testid={`${populateId()}table_row_${rowIndex}_cell_${cellIndex}`}
                    key={cellIndex.toString()}
                    cell={cell}
                    onCellClick={handleCellClick}
                    {...cellProps}
                  >
                    {cell.render('Cell')}
                  </TableCell>
                ))}
              </TableRow>
            );
          })}
        </tbody>
      </RsTable>
    </TableContext.Provider>
  );
};

Table.propTypes = {
  id: PropTypes.string,
  additionalContent: Util.tagPropType,
  bodyProps: PropTypes.object,
  cellProps: PropTypes.object,
  columns: PropTypes.arrayOf(PropTypes.object),
  headerProps: PropTypes.object,
  initialState: PropTypes.object,
  onRowClick: PropTypes.func,
  onRowSelected: PropTypes.func,
  selectable: PropTypes.bool,
  sortable: PropTypes.bool,
  scrollable: PropTypes.bool,
  records: PropTypes.arrayOf(PropTypes.object),
  rowProps: PropTypes.object,
  sortBy: PropTypes.arrayOf(PropTypes.object),
  getToggleAllRowsSelectedProps: PropTypes.func,
  row: PropTypes.object,
};

Table.defaultProps = {
  headerProps: {},
  rowProps: {},
  bodyProps: {},
  cellProps: {},
};

export default Table;
