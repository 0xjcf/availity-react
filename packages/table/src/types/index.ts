export type { OnTableClickEvent } from './OnTableClickEvent';
export type {
    OnRowSelectedEvent
} from './OnRowSelectedEvent';

export type {
    Row,
    Column,
    Cell,
    ExtendedTableHeader,
    CurrentTableState,
    TableInstance
} from './ReactTable';

export type {
    PrimaryTableAction,
    SingleTableAction,
    TableAction,
    BulkTableAction
} from './TableActions';

export type {
    TableSortConfig
} from './TableSortConfig';

export type {
    TableSortOption
} from './TableSortOption';