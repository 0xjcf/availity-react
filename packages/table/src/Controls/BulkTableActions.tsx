import React, { useEffect, useState } from 'react';
import { Badge, Button, ButtonDropdown, ButtonGroup, DropdownItem, DropdownMenu, DropdownToggle } from 'reactstrap';
import { useTableContext } from '../TableContext';
import { TableInstance } from '../types/ReactTable';
import { BulkTableAction } from '../types/TableActions';
import { TableRecord } from '../types/TableRecord';

type Props = {
    id?: string;
    disabled?: boolean;
    recordName: string;
    bulkActions: BulkTableAction[];
    color?: string;
    onRecordsSelected?: (records: TableRecord[]) => void;
} & React.HTMLAttributes<HTMLElement>;

const BulkTableActions = ({ id, disabled, color, recordName, bulkActions, onRecordsSelected }: Props): JSX.Element | null => {
    const { instance } = useTableContext();

    const { selectedFlatRows: selectedRows, isAllRowsSelected, toggleAllRowsSelected } = instance as TableInstance;

    const [isSelectionDropdownOpen, setIsSelectionDropdownOpen] = useState<boolean>(false);
    const [numberOfSelectedRows, setNumberOfSelectedRows] = useState<number>(0);
    const [selectionButtonText, setSelectionButtonText] = useState('Select');
    const [isDisabled, setIsDisabled] = useState<boolean>(disabled || false);

    useEffect(() => {
        setNumberOfSelectedRows(selectedRows?.length);
        if (isAllRowsSelected) {
            setSelectionButtonText('Deselect');
        } else {
            setSelectionButtonText('Select');
        }

        if (onRecordsSelected) {
            const records = selectedRows.map(row =>  row.original);
            onRecordsSelected(records);
        }

        setIsDisabled(disabled || selectedRows?.length === 0);
    }, [selectedRows, disabled, numberOfSelectedRows]);

    const toggleSelectionDropdown = () => setIsSelectionDropdownOpen(!isSelectionDropdownOpen);

    const handleToggleSelectAll = () => {
        if (toggleAllRowsSelected) {
            toggleAllRowsSelected();
        }
    }

    return (
        <ButtonGroup id={id} disabled={isDisabled} className="btn-group">
            <Button
                disabled={isDisabled}
                onClick={() => { handleToggleSelectAll() }}
                color={color}>
                <Badge>{numberOfSelectedRows}</Badge> {selectionButtonText} All {recordName}
            </Button>
            <ButtonDropdown disabled={isDisabled} isOpen={isSelectionDropdownOpen} toggle={toggleSelectionDropdown}>
                <DropdownToggle disabled={isDisabled} color={color} caret />
                <DropdownMenu color={color}>
                    {bulkActions.map((action) => {
                        const isVisible = action.isVisible ? action.isVisible() : true;
                        const setProps = () => {
                            if (!action.onClick) {
                                return null;
                            }
                            const clickEvent = action.onClick;
                            return { onClick: () => clickEvent(selectedRows.map(row =>  row.original)) };
                        }

                        if (isVisible) {
                            return <DropdownItem key={action.displayText} {...setProps()}>{action.displayText}</DropdownItem>
                        }
                        return null;
                    }
                    )}
                </DropdownMenu>
            </ButtonDropdown>
        </ButtonGroup>
    );
}

export default BulkTableActions;