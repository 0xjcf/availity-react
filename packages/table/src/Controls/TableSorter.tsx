import React, { useEffect, useState } from 'react';
import {
  Button,
  ButtonDropdown,
  DropdownItem,
  DropdownMenu,
  DropdownToggle,
  InputGroup,
  Label,
  UncontrolledTooltip,
} from 'reactstrap';
import Icon from '@availity/icon';
import find from 'lodash/find';
import { TableSort } from '../types/TableSort';
import { useTableContext } from '../TableContext';
import { CurrentTableState, IdType, TableInstance } from '../types/ReactTable';
import { TableSortOption } from '../types/TableSortOption';

type Props = {
  id?: string;
  disabled?: boolean;
  color?: string;
  onSort?: (sortBy: TableSort[]) => void;
  sortOptions?: TableSortOption[];
  autoGenerateSortOptions?: boolean;
} & React.HTMLAttributes<HTMLElement>;

const TableSorter = <T extends IdType>({ id, onSort, disabled, color, sortOptions, autoGenerateSortOptions, ...rest }: Props): JSX.Element | null => {
  const { sortableColumns, instance } = useTableContext();
  const { toggleSortBy, state } = instance as TableInstance<T>;

  const [isSortingDropdownOpen, setIsSortingDropdownOpen] = useState<boolean>(false);
  const [tableSort, setTableSort] = useState<TableSortOption | undefined>();
  const [tableSortDesc, setTableSortDesc] = useState<boolean>(true);
  const [isDisabled, setIsDisabled] = useState<boolean>(disabled || false);
  const [displayedSortOptions, setDisplayedSortOptions] = useState<TableSortOption[]>();

  useEffect(() => {
    if (sortableColumns) {
      let availableSortOptions: TableSortOption[] = [];
      if (autoGenerateSortOptions) {
        availableSortOptions = sortableColumns;
      }

      if (sortOptions) {
        availableSortOptions = [...availableSortOptions, ...sortOptions];
      }

      setDisplayedSortOptions(availableSortOptions);
    }
  }, [sortOptions, autoGenerateSortOptions, sortableColumns]);

  useEffect(() => {
    if (state && displayedSortOptions) {
      const { sortBy } = state as CurrentTableState;
      if (!sortBy) {
        return;
      }

      const currentSort = (sortBy as TableSort[])[0];
      if (!currentSort) {
        return;
      }

      const currentSortOption = find(displayedSortOptions, (option) => option.value === currentSort.id);
      if (currentSortOption) {
        setTableSort(currentSortOption);
        setTableSortDesc(currentSort.desc);
      }
    }
  }, [state, displayedSortOptions])

  useEffect(() => {
    setIsDisabled(disabled || false);
  }, [disabled]);

  const sort = async (sortBy: TableSortOption, isDesc: boolean) => {
    if (toggleSortBy) {
      toggleSortBy(sortBy.value, isDesc, false);
    }

    if (onSort) {
      onSort([{ id: sortBy.value, desc: isDesc }]);
    }
  };

  const toggleSortingDropdown = () => setIsSortingDropdownOpen(!isSortingDropdownOpen);

  const handleSort = async (sortOption: TableSortOption) => {
    await setTableSort(sortOption);

    let isDesc = tableSortDesc;
    if (sortOption.isDesc !== undefined) {
      isDesc = sortOption.isDesc;
      setTableSortDesc(isDesc);
    }

    sort(sortOption, isDesc);
  };

  const toggleSortDirection = async () => {
    const newSortDir = !tableSortDesc;
    await setTableSortDesc(newSortDir);

    if (!tableSort) {
      return;
    }

    sort(tableSort, newSortDir);
  };

  if (!displayedSortOptions) {
    return null;
  }

  return (
    <InputGroup
      data-testid="sorter_input_group"
      id={id}
      className="table-controls-sorting mb-2 me-sm-2 mb-sm-0"
      {...rest}
    >
      <Label className="me-sm-2 pr-1 pl-2">
        <strong>Sort By: </strong>
      </Label>
      <ButtonDropdown className="m1-1" isOpen={isSortingDropdownOpen} toggle={toggleSortingDropdown}>
        <DropdownToggle data-testid="sorter_toggle" disabled={isDisabled} color={color} caret>
          {tableSort && tableSort.label}
        </DropdownToggle>
        <DropdownMenu data-testid="sorter_menu" disabled={isDisabled} color={color}>
          {displayedSortOptions?.map((sortOption: TableSortOption) => (
            <DropdownItem
              data-testid={`sorter_menu_${sortOption.value}`}
              key={sortOption.value}
              onClick={() => handleSort(sortOption)}
            >
              {sortOption.label}
            </DropdownItem>
          ))}
        </DropdownMenu>
        <UncontrolledTooltip
          data-testid="tooltip_toggle_sort_dir"
          id="tooltip-toggle-sort-dir"
          placement="top"
          target="btn-toggle-sort-dir"
        >
          {tableSortDesc ? 'Descending' : 'Ascending'}
        </UncontrolledTooltip>
        <Button
          data-testid="btn_toggle_sort_dir"
          id="btn-toggle-sort-dir"
          disabled={isDisabled}
          color={color}
          onClick={async () => {
            await toggleSortDirection();
          }}
        >
          <Icon data-testid="btn_toggle_sort_dir_icon" name={tableSortDesc ? 'sort-alt-down' : 'sort-alt-up'} />
        </Button>
      </ButtonDropdown>
    </InputGroup>
  );
};

TableSorter.defaultProps = {
  autoGenerateSortOptions: true
}
export default TableSorter;
