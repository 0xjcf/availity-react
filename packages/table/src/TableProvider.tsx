import React, { useEffect, useState } from 'react';
import { useRowSelect, useSortBy, useTable, Hooks, UseRowSelectInstanceProps, Column as RtColumn } from 'react-table';
import filter from 'lodash/filter';
import IndeterminateCheckbox from './IndeterminateCheckbox';
import { Cell, Column, CurrentTableState, IdType, TableInstance, TableOptions } from './types/ReactTable';
import { TableSortOption } from './types/TableSortOption';
import { TableContext } from './TableContext';

export type TableProviderProps<T extends IdType> = {
  additionalContent?: React.ElementType;
  additionalContentProps?: Record<string, string | number | boolean | undefined | null>;

  children: React.ReactChild | React.ReactChild[];
  columns: Column<T>[];
  data: T[];
  scrollable?: boolean;
  selectable?: boolean;
  sortable?: boolean;
  initialState?: Partial<CurrentTableState>;
} & TableOptions<T>;

const TableProvider = <T extends IdType>({
  additionalContent: AdditionalContent,
  additionalContentProps,

  columns,
  data,
  selectable,
  scrollable,
  sortable,
  children,
  ...rest
}: TableProviderProps<T>): JSX.Element => {
  let selectionColumn: Column<T>;
  const [isScrollable, setScrollable] = useState<boolean | undefined>(scrollable);

  const getSortableColumns = (): TableSortOption[] =>
    filter(columns, (column) => !column.disableSortBy && column.defaultCanSort).map((column) => {
      const col = column as Column<T>;
      return { value: col.accessor as string, label: col.Header as string };
    });

  const cols = columns as RtColumn<T>[];

  const tableInstance = useTable<T>(
    {
      ...rest,
      columns: cols,
      data,
    } as TableOptions<T>,
    useSortBy,
    useRowSelect,
    (hooks: Hooks<T>) => {
      selectionColumn = {
        id: 'selection',
        title: 'Select record(s)',
        className: 'fixed-width-selection',
        defaultCanSort: false,
        disableSortBy: true,
        disableClick: true,
        Header: ({ getToggleAllRowsSelectedProps }: UseRowSelectInstanceProps<T>) => (
          <div className="text-center">
            <IndeterminateCheckbox
              data-testid="table_header_select_all"
              aria-label="Select all records"
              {...getToggleAllRowsSelectedProps()}
            />
          </div>
        ),
        Cell: ({ row: { getToggleRowSelectedProps, index } }: Cell<T>) => (
          <div className="text-center">
            <IndeterminateCheckbox
              data-testid={`table_header_select_row_${index}`}
              aria-label="Select record"
              {...getToggleRowSelectedProps()}
            />
          </div>
        ),
      };

      hooks.visibleColumns.push((columns: Column<T>[]) => [selectionColumn, ...columns]);
    }
  ) as TableInstance<T>;

  const { toggleHideColumn } = tableInstance;

  useEffect(() => {
    toggleHideColumn('selection', !selectable);
  }, [selectable, toggleHideColumn]);

  return (
    <TableContext.Provider
      value={{
        AdditionalContent,
        additionalContentProps,
        selectable,
        scrollable: isScrollable as boolean | undefined,
        setScrollable,
        sortable,
        sortableColumns: getSortableColumns(),
        instance: tableInstance,
      }}
    >
      {children}
    </TableContext.Provider>
  );
};

export default TableProvider;
