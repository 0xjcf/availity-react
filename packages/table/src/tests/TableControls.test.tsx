import React from 'react';
import { render, waitFor } from '@testing-library/react';
import every from 'lodash/every';
import basicData from './data/basicData.json';
import TableControls from '../Controls/TableControls';
import BulkTableActions from '../Controls/BulkTableActions';
import TableProvider from '../TableProvider';
import { Column } from '../types/ReactTable';
import TableSorter from '../Controls/TableSorter';
import Table from '../Table';

const basicColumns = [
  {
    Header: 'First Name',
    accessor: 'first_name',
    defaultCanSort: true,
    disableSortBy: false,
  },
  {
    Header: 'Last Name',
    accessor: 'last_name',
    defaultCanSort: false,
    disableSortBy: true,
  },
  {
    Header: 'Email',
    accessor: 'email',
  },
] as Column<Record<string, string | boolean | number>>[];

const bulkActions = [
  {
    id: 'action1',
    displayText: 'Action 1',
    isVisible: (records?: Record<string, string | boolean | number>[]) => every(records, { isActive: true }),
    onClick: (records?: Record<string, string | boolean | number>[]) => {
      if (!records) {
        return;
      }

      for (const record of records) {
        // eslint-disable-next-line no-console
        console.log(`action on record ${record?.id}`);
      }
    },
  },
  {
    id: 'action2',
    displayText: 'Action 2',
    onClick: (records?: Record<string, string>[]) => {
      if (!records) {
        return;
      }

      for (const record of records) {
        // eslint-disable-next-line no-console
        console.log(`action on record ${record?.id}`);
      }
    },
  },
];

describe('TableControls', () => {
  test('should render table controls with one child', () => {
    const { container } = render(
      <TableProvider data={basicData} columns={basicColumns}>
        <TableControls>
          <BulkTableActions bulkActions={bulkActions} />
        </TableControls>
      </TableProvider>
    );

    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  test('should render table controls with two children', () => {
    const { container } = render(
      <TableProvider>
        <TableControls>
          <BulkTableActions bulkActions={bulkActions} />
          <TableSorter />
        </TableControls>
        <Table data={basicData} columns={basicColumns}></Table>

      </TableProvider>
    );

    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  test('should disable all records when disabled is true', async () => {
    const { container, getByTestId } = render(
      <TableProvider>
        <TableControls disabled>
          <BulkTableActions bulkActions={bulkActions} />
          <TableSorter />
        </TableControls>
        <Table data={basicData} columns={basicColumns}></Table>

      </TableProvider>
    );

    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();

    const bulkActionsBtn = await waitFor(() => getByTestId('bulk_actions_btn_group'));
    expect(bulkActionsBtn).not.toBeNull();
    expect(bulkActionsBtn).toHaveAttribute('disabled');

    const sorterInput = await waitFor(() => getByTestId('sorter_toggle'));
    expect(sorterInput).not.toBeNull();
    expect(sorterInput).toHaveAttribute('disabled');
  });

  test('should disable all records when disabled is false', async () => {
    const { container, getByTestId } = render(
      <TableProvider>
        <TableControls>
          <BulkTableActions bulkActions={bulkActions} />
          <TableSorter />
        </TableControls>
        <Table data={basicData} columns={basicColumns}></Table>
      </TableProvider>
    );

    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();

    const bulkActionsBtn = await waitFor(() => getByTestId('bulk_actions_btn_group'));
    expect(bulkActionsBtn).not.toBeNull();
    expect(bulkActionsBtn).not.toHaveAttribute('disabled');

    const sorterInput = await waitFor(() => getByTestId('sorter_toggle'));
    expect(sorterInput).not.toBeNull();
    expect(sorterInput).not.toHaveAttribute('disabled');
  });
});
