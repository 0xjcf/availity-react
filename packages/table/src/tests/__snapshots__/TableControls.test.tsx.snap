// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableControls should disable all records when disabled is false 1`] = `
<div>
  <div
    class="table-controls"
    data-testid="test-controls-container"
  >
    <form
      class="form-inline"
    >
      <div
        class="btn-group btn-group"
        data-testid="bulk_actions_btn_group"
        role="group"
      >
        <button
          class="btn btn-secondary"
          data-testid="select_deselect_all_records"
          type="button"
        >
          <span
            class="badge badge-secondary badge-pill"
          >
            0
          </span>
           
          Select
           All 
          Records
        </button>
        <div
          class="btn-group"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-secondary disabled"
            data-testid="bulk_actions_toggle"
            disabled=""
            type="button"
          >
            <span
              class="sr-only"
            >
              Toggle Dropdown
            </span>
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item"
              data-testid="bulk_action_action1"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Action 1
            </button>
            <button
              class="dropdown-item"
              data-testid="bulk_action_action2"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Action 2
            </button>
          </div>
        </div>
      </div>
      <div
        class="table-controls-sorting mb-2 me-sm-2 mb-sm-0 input-group"
        data-testid="sorter_input_group"
      >
        <label
          class="me-sm-2 pr-1 pl-2"
        >
          <strong>
            Sort By: 
          </strong>
        </label>
        <div
          class="m1-1 btn-group"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-secondary"
            data-testid="sorter_toggle"
            type="button"
          >
            <span
              class="sr-only"
            >
              Toggle Dropdown
            </span>
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu"
            data-testid="sorter_menu"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item"
              data-testid="sorter_menu_first_name"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              First Name
            </button>
          </div>
          <button
            class="btn btn-secondary"
            data-testid="btn_toggle_sort_dir"
            id="btn-toggle-sort-dir"
            type="button"
          >
            <i
              aria-hidden="true"
              class="icon icon-sort-alt-down"
              data-testid="btn_toggle_sort_dir_icon"
            />
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`TableControls should disable all records when disabled is true 1`] = `
<div>
  <div
    class="table-controls"
    data-testid="test-controls-container"
  >
    <form
      class="form-inline"
    >
      <div
        class="btn-group btn-group"
        data-testid="bulk_actions_btn_group"
        disabled=""
        role="group"
      >
        <button
          class="btn btn-secondary disabled"
          data-testid="select_deselect_all_records"
          disabled=""
          type="button"
        >
          <span
            class="badge badge-secondary badge-pill"
          >
            0
          </span>
           
          Select
           All 
          Records
        </button>
        <div
          class="btn-group"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-secondary disabled"
            data-testid="bulk_actions_toggle"
            disabled=""
            type="button"
          >
            <span
              class="sr-only"
            >
              Toggle Dropdown
            </span>
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item"
              data-testid="bulk_action_action1"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Action 1
            </button>
            <button
              class="dropdown-item"
              data-testid="bulk_action_action2"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Action 2
            </button>
          </div>
        </div>
      </div>
      <div
        class="table-controls-sorting mb-2 me-sm-2 mb-sm-0 input-group"
        data-testid="sorter_input_group"
      >
        <label
          class="me-sm-2 pr-1 pl-2"
        >
          <strong>
            Sort By: 
          </strong>
        </label>
        <div
          class="m1-1 btn-group"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-secondary disabled"
            data-testid="sorter_toggle"
            disabled=""
            type="button"
          >
            <span
              class="sr-only"
            >
              Toggle Dropdown
            </span>
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu"
            data-testid="sorter_menu"
            disabled=""
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item"
              data-testid="sorter_menu_first_name"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              First Name
            </button>
          </div>
          <button
            class="btn btn-secondary disabled"
            data-testid="btn_toggle_sort_dir"
            disabled=""
            id="btn-toggle-sort-dir"
            type="button"
          >
            <i
              aria-hidden="true"
              class="icon icon-sort-alt-down"
              data-testid="btn_toggle_sort_dir_icon"
            />
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`TableControls should render table controls with one child 1`] = `
<div>
  <div
    class="table-controls"
    data-testid="test-controls-container"
  >
    <form
      class="form-inline"
    >
      <div
        class="btn-group btn-group"
        data-testid="bulk_actions_btn_group"
        role="group"
      >
        <button
          class="btn btn-secondary"
          data-testid="select_deselect_all_records"
          type="button"
        >
          <span
            class="badge badge-secondary badge-pill"
          >
            0
          </span>
           
          Select
           All 
          Records
        </button>
        <div
          class="btn-group"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-secondary disabled"
            data-testid="bulk_actions_toggle"
            disabled=""
            type="button"
          >
            <span
              class="sr-only"
            >
              Toggle Dropdown
            </span>
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item"
              data-testid="bulk_action_action1"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Action 1
            </button>
            <button
              class="dropdown-item"
              data-testid="bulk_action_action2"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Action 2
            </button>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`TableControls should render table controls with two children 1`] = `
<div>
  <div
    class="table-controls"
    data-testid="test-controls-container"
  >
    <form
      class="form-inline"
    >
      <div
        class="btn-group btn-group"
        data-testid="bulk_actions_btn_group"
        role="group"
      >
        <button
          class="btn btn-secondary"
          data-testid="select_deselect_all_records"
          type="button"
        >
          <span
            class="badge badge-secondary badge-pill"
          >
            0
          </span>
           
          Select
           All 
          Records
        </button>
        <div
          class="btn-group"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-secondary disabled"
            data-testid="bulk_actions_toggle"
            disabled=""
            type="button"
          >
            <span
              class="sr-only"
            >
              Toggle Dropdown
            </span>
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item"
              data-testid="bulk_action_action1"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Action 1
            </button>
            <button
              class="dropdown-item"
              data-testid="bulk_action_action2"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              Action 2
            </button>
          </div>
        </div>
      </div>
      <div
        class="table-controls-sorting mb-2 me-sm-2 mb-sm-0 input-group"
        data-testid="sorter_input_group"
      >
        <label
          class="me-sm-2 pr-1 pl-2"
        >
          <strong>
            Sort By: 
          </strong>
        </label>
        <div
          class="m1-1 btn-group"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-secondary"
            data-testid="sorter_toggle"
            type="button"
          >
            <span
              class="sr-only"
            >
              Toggle Dropdown
            </span>
          </button>
          <div
            aria-hidden="true"
            class="dropdown-menu"
            data-testid="sorter_menu"
            role="menu"
            tabindex="-1"
          >
            <button
              class="dropdown-item"
              data-testid="sorter_menu_first_name"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              First Name
            </button>
          </div>
          <button
            class="btn btn-secondary"
            data-testid="btn_toggle_sort_dir"
            id="btn-toggle-sort-dir"
            type="button"
          >
            <i
              aria-hidden="true"
              class="icon icon-sort-alt-down"
              data-testid="btn_toggle_sort_dir_icon"
            />
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
